src/Polysemy/MTL.hs

```diff
@@ -0,0 +1,189 @@
+  
+------------------------------------------------------------------------------
+absorbState :: forall s r a. Member (State s) r => (S.MonadState s (Sem r) => Sem r a) -> Sem r a
+absorbState f = absorb @(S.MonadState s) f -- using (MonadState get put) f
+
+instance ReifiableConstraint1 (S.MonadState s) where
+  data Dict (S.MonadState s) m = MonadState
+    { get_ :: m s
+    , put_ :: s -> m ()
+    }
+  reifiedInstance = C.Sub C.Dict
+
+instance ( Monad m
+         , R.Reifies s' (Dict (S.MonadState s) m)
+         ) => S.MonadState s (ConstrainedAction (S.MonadState s) m s') where
+  get = ConstrainedAction $ get_ $ R.reflect $ Proxy @s'
+  put s = ConstrainedAction $ put_ (R.reflect $ Proxy @s') s
+
+
+instance Member (State s) r => SemDictionary (S.MonadState s) r where
+  semDictionary = MonadState get put
+
+--------------------------------------------------------------------------------
+absorbWriter :: forall w r a. (Monoid w, Member (Writer w) r) => (S.MonadWriter w (Sem r) => Sem r a) -> Sem r a
+absorbWriter = absorb @(S.MonadWriter w)
+
+instance Monoid w => ReifiableConstraint1 (S.MonadWriter w) where
+  data Dict (S.MonadWriter w) m = MonadWriter
+    { tell_ :: w -> m ()
+    , listen_ :: forall a. m a -> m (a, w)
+    , pass_ :: forall a. m (a, w -> w) -> m a 
+    }
+  reifiedInstance = C.Sub C.Dict
+
+instance ( Monad m
+         , Monoid w
+         , R.Reifies s' (Dict (S.MonadWriter w) m)
+         ) => S.MonadWriter w (ConstrainedAction (S.MonadWriter w) m s') where
+  tell w = ConstrainedAction $ tell_ (R.reflect $ Proxy @s') w
+  listen x = ConstrainedAction $ listen_ (R.reflect $ Proxy @s') (action x)
+  pass x = ConstrainedAction $ pass_ (R.reflect $ Proxy @s') (action x)
+
+{- This one requires a little work since the polysemy writer is a bit different from the
+mtl-standard one
+-} 
+instance (Monoid w, Member (Writer w) r) => SemDictionary (S.MonadWriter w) r where
+  semDictionary = MonadWriter tell semListen semPass where
+    semListen = fmap (\(x,y) -> (y,x)) . listen
+    semPass :: Member (Writer w) r => Sem r (a, w -> w) -> Sem r a 
+    semPass x = do
+      (a, f) <- x
+      censor f (return a)
+
+--------------------------------------------------------------------------------
+absorbError :: forall e r a. Member (Error e) r => (S.MonadError e (Sem r) => Sem r a) -> Sem r a
+absorbError = absorb @(S.MonadError e)
+
+instance ReifiableConstraint1 (S.MonadError e) where
+  data Dict (S.MonadError e) m = MonadError
+    { throwError_ :: forall a. e -> m a
+    , catchError_ :: forall a. m a -> (e -> m a) -> m a
+    }
+  reifiedInstance = C.Sub C.Dict
+
+instance ( Monad m
+         , R.Reifies s' (Dict (S.MonadError e) m)
+         ) => S.MonadError e (ConstrainedAction (S.MonadError e) m s') where
+  throwError e = ConstrainedAction $ throwError_ (R.reflect $ Proxy @s') e
+  catchError x f = ConstrainedAction $ catchError_ (R.reflect $ Proxy @s') (action x) (action . f)
```

Comment:
Please add `INLINEABLE` pragmas to every class mtl class method here, and to every `semDictionary` definition.